### PR TITLE
feat(app): Auth code flow issuance ui

### PR DESCRIPTION
**Pull request by talwinder50**
In order to test this flow you need real device as launching the url in the browser does not work in simulator. 
Also you would need to replace the client and redirect uri with the value - till 
I push the follow up pr to read the clientID and uri from the flutter environment variable. #423 
Also for error handling from the sdk #426 will be pushed as follow up pr

Signed-off-by: Talwinder kaur <talwinder.kaur@avast.com>

### CODE REVIEW

**Review by birtony — APPROVED** [review]

File: demo/app/ios/Runner/OpenID4CI.swift [hunk 1]
```diff
@@ -31,14 +31,53 @@ public class OpenID4CI {
...
+        
+        return authorizationLink
+    }
+
+    
+    
     func pinRequired() throws -> Bool {
+        let issuerCapabilities = initiatedInteraction.issuerCapabilities()
+        if  !issuerCapabilities!.preAuthorizedCodeGrantTypeSupported() {
+            // This code example isn't applicable. See the authorization code flow example instead.
```
Comment (DRK3) [inline demo/app/ios/Runner/OpenID4CI.swift:65]: Should remove this comment

File: demo/app/ios/Runner/OpenID4CI.swift [hunk 2]
```diff
@@ -31,14 +31,53 @@ public class OpenID4CI {
...
+    
+    func getAuthorizationLink() throws -> String {
+        let issuerCapabilities = initiatedInteraction.issuerCapabilities()
+        if !(issuerCapabilities?.authorizationCodeGrantTypeSupported())! {
+            // This code example isn't applicable. See the authorization code flow example instead.
+            return "Not implemented"
+        }
+        
+        let scopes = ApiStringArray()
+        scopes!.append("openid")!.append("profile")
```
Comment (DRK3) [inline demo/app/ios/Runner/OpenID4CI.swift:53]: Should leave these blank for now

File: demo/app/ios/Runner/OpenID4CI.swift [hunk 3]
```diff
@@ -31,14 +31,53 @@ public class OpenID4CI {
...
+        if ((issuerCapabilities!.preAuthorizedCodeGrantTypeSupported())){
+           return "preauth-code-flow"
+        }
+        return ""
+    }
+    
+    func getAuthorizationLink() throws -> String {
+        let issuerCapabilities = initiatedInteraction.issuerCapabilities()
+        if !(issuerCapabilities?.authorizationCodeGrantTypeSupported())! {
+            // This code example isn't applicable. See the authorization code flow example instead.
```
Comment (DRK3) [inline demo/app/ios/Runner/OpenID4CI.swift:48]: Should remove this comment (doesn't make sense)

File: demo/app/ios/Runner/OpenID4CI.swift [hunk 4]
```diff
@@ -31,14 +31,53 @@ public class OpenID4CI {
...
+            return "Not implemented"
```
Comment (DRK3) [inline demo/app/ios/Runner/OpenID4CI.swift:49]: Should throw an error

File: demo/app/ios/Runner/OpenID4CI.swift [hunk 5]
```diff
@@ -31,14 +31,53 @@ public class OpenID4CI {
...
+    }
+
+    
+    
     func pinRequired() throws -> Bool {
+        let issuerCapabilities = initiatedInteraction.issuerCapabilities()
+        if  !issuerCapabilities!.preAuthorizedCodeGrantTypeSupported() {
+            // This code example isn't applicable. See the authorization code flow example instead.
+            
+            return false
```
Comment (DRK3) [inline demo/app/ios/Runner/OpenID4CI.swift:67]: Need to throw an error here

File: demo/app/ios/Runner/OpenID4CI.swift [hunk 6]
```diff
@@ -31,14 +31,53 @@ public class OpenID4CI {
...
+        let issuerCapabilities = initiatedInteraction.issuerCapabilities()
+        if !(issuerCapabilities?.authorizationCodeGrantTypeSupported())! {
+            // This code example isn't applicable. See the authorization code flow example instead.
+            return "Not implemented"
+        }
+        
+        let scopes = ApiStringArray()
+        scopes!.append("openid")!.append("profile")
+        // TODO #423 Read withScopes and redirect uri from flutter enviornment. Replace these with approriate values as of now.
+        let authorizationLink = initiatedInteraction.createAuthorizationURL(withScopes: "clientID", redirectURI: "redirect URI", scopes: scopes, error: nil)
```
Comment (DRK3) [inline demo/app/ios/Runner/OpenID4CI.swift:55]: Need to ensure no error here

File: demo/app/ios/Runner/OpenID4CI.swift (hunk 5 above)
Comment (DRK3) [reply demo/app/ios/Runner/OpenID4CI.swift:67]: (if pre-auth code flow isn't support, then `pinRequired` doesn't make sense)

File: demo/app/ios/Runner/flutterPlugin.swift [hunk 7]
```diff
@@ -286,7 +290,7 @@ public class SwiftWalletSDKPlugin: NSObject, FlutterPlugin {
      AuthorizeResult is the object returned from the Client.Authorize method.
```
Comment (DRK3) [inline demo/app/ios/Runner/flutterPlugin.swift:290]: comment is out of date

File: demo/app/lib/scenarios/launched_external_url.dart [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+                MaterialPageRoute(builder: (context) => RedirectPathTextBox()),
+              );
+            });
+          }
+          return CircularProgressIndicator();
+        },
+      ),
+    );
+  }
+}
```
Comment (DRK3) [inline demo/app/lib/scenarios/launched_external_url.dart:40]: Missing end of file newline

**Comment by DRK3** [comment]
> In order to test this flow you need real device as launching the url in the browser does not work in simulator. Also you would need to replace the client and redirect uri with the value - till I push the follow up pr to read the client id and uri from the flutter environment variable. #423
> 
> Signed-off-by: Talwinder kaur [talwinder.kaur@avast.com](mailto:talwinder.kaur@avast.com)

Should say `clientID` instead of just `client`. Also, the user of the app would need to add in their own scopes as well as the client ID and redirect URI.

File: demo/app/ios/Runner/OpenID4CI.swift (hunk 4 above)
Comment (talwinder50) [reply demo/app/ios/Runner/OpenID4CI.swift:49]: error handling I am doing separately in this pr #426 as I have to catch properly in the flutter layer and show error too

File: demo/app/ios/Runner/OpenID4CI.swift (hunk 5 above)
Comment (talwinder50) [reply demo/app/ios/Runner/OpenID4CI.swift:67]: error handling I am doing separately in this pr https://github.com/trustbloc/wallet-sdk/issues/426 as I have to catch properly in the flutter layer and show error too

File: demo/app/ios/Runner/OpenID4CI.swift (hunk 6 above)
Comment (talwinder50) [reply demo/app/ios/Runner/OpenID4CI.swift:55]: same as above

**Review by DRK3 — APPROVED** [review]

**Pull request merged**